### PR TITLE
Support references in InfoItem

### PR DESCRIPTION
**Pull request by huyhg**
Implements proposal in #12.

### CODE REVIEW

File: hack/install.yaml [hunk 1]
```diff
@@ -68,30 +68,45 @@ spec:
...
+                  valueFrom:
+                    properties:
+                      ingressRef:
+                        properties:
+                          host:
+                            type: string
+                          path:
+                            type: string
+                        type: object
+                      secretRef:
```
Comment (prydonius) [inline hack/install.yaml]: I think this should be `secretKeyRef` to match what's used in the podspec. Also there should be a `configMapKeyRef` too, right?

File: pkg/apis/app/v1alpha1/application_types.go [hunk 2]
```diff
@@ -100,6 +101,42 @@ type InfoItem struct {
 
 	// Value is human readable content.
 	Value string `json:"value,omitempty"`
+
+	// Value is derived from another source.
+	ValueFrom *InfoItemSource `json:"valueFrom,omitempty"`
+}
+
+type InfoItemSource struct {
+  SecretKeyRef *ConfigMapKeySelector `json:"secretRef,omitempty"`
```
Comment (prydonius) [inline pkg/apis/app/v1alpha1/application_types.go]: could we not just reuse `core.SecretKeySelector`?

File: hack/install.yaml (hunk 1 above)
Comment (huyhg) [reply hack/install.yaml]: Good catch, thanks! I missed updating the json annotation.

File: pkg/apis/app/v1alpha1/application_types.go (hunk 2 above)
Comment (huyhg) [reply pkg/apis/app/v1alpha1/application_types.go]: Thanks for the pointer. Yes, we can re-use it. Updated!

File: hack/install.yaml [hunk 3]
```diff
@@ -68,30 +68,42 @@ spec:
...
+                      configMapKeyRef:
+                        type: object
+                      ingressRef:
+                        properties:
+                          host:
+                            type: string
+                          path:
+                            type: string
+                        type: object
+                      secretKeyRef:
```
Comment (huyhg) [inline hack/install.yaml]: @kow3ns I'm using the latest kubebuilder, but it doesn't seem to be able to pull in the definitions of core structs here.

File: hack/install.yaml [hunk 4]
```diff
@@ -186,9 +182,50 @@ metadata:
...
+        resources:
+          limits:
+            cpu: 100m
+            memory: 30Mi
+          requests:
+            cpu: 100m
+            memory: 20Mi
+      terminationGracePeriodSeconds: 10
+  updateStrategy: {}
+status:
```
Comment (huyhg) [inline hack/install.yaml]: @kow3ns not sure what's going on with the generated code here...

**Comment by huyhg** [comment]
@konryd FYI

File: pkg/apis/app/v1alpha1/application_types.go [hunk 5]
```diff
@@ -100,6 +101,34 @@ type InfoItem struct {
 
 	// Value is human readable content.
 	Value string `json:"value,omitempty"`
+
```
Comment (kow3ns) [inline pkg/apis/app/v1alpha1/application_types.go]: You are implementing the golang version of a union. You need to add a type specified to indicate whether Value or ValueFrom is set e.g. 
```go
type InfoItemType string

const (
  Value InfoItemType = "value"
  Reference = "reference"
)
```
You may also want to make Value a pointer. Only one should be set, and it should correspond to the indicated type field.

File: pkg/apis/app/v1alpha1/application_types.go [hunk 6]
```diff
@@ -100,6 +101,34 @@ type InfoItem struct {
...
+type InfoItemSource struct {
+	SecretKeyRef    *core.SecretKeySelector    `json:"secretKeyRef,omitempty"`
+	ConfigMapKeyRef *core.ConfigMapKeySelector `json:"configMapKeyRef,omitempty"`
+	ServiceRef      *ServiceSelector           `json:"serviceRef,omitempty"`
+	IngressRef      *IngressSelector           `json:"ingressRef,omitempty"`
+}
+
+type ServiceSelector struct {
+	// The Service to select from.
+	core.LocalObjectReference `json:",inline"`
```
Comment (kow3ns) [inline pkg/apis/app/v1alpha1/application_types.go]: If you use LocalObjectReference the Service must appear in the same namespace as the Application. The same is true of Ingress. Is this only meant to work with load balanced services?

File: pkg/apis/app/v1alpha1/application_types.go [hunk 7]
```diff
@@ -100,6 +101,34 @@ type InfoItem struct {
 
 	// Value is human readable content.
 	Value string `json:"value,omitempty"`
+
+	// Value is derived from another source.
+	ValueFrom *InfoItemSource `json:"valueFrom,omitempty"`
+}
+
+type InfoItemSource struct {
```
Comment (kow3ns) [inline pkg/apis/app/v1alpha1/application_types.go]: InfoItemSource, ServiceSelector, and IngressSelector need comments.

File: pkg/apis/app/v1alpha1/application_types.go [hunk 8]
```diff
@@ -100,6 +101,34 @@ type InfoItem struct {
...
+	SecretKeyRef    *core.SecretKeySelector    `json:"secretKeyRef,omitempty"`
```
Comment (kow3ns) [inline pkg/apis/app/v1alpha1/application_types.go]: You want need to have a type field to indicate the member of the union that contains the value. The same as InfoItem above. Each of the values needs a comment to indicate the use.

File: pkg/apis/app/v1alpha1/application_types.go (hunk 5 above)
Comment (huyhg) [reply pkg/apis/app/v1alpha1/application_types.go]: Understood your comment about union in Go. Such type wouldn't be consistent with what is currently in the core API though. I'm drawing from the following examples:

https://github.com/kubernetes/api/blob/184e700b32b7f1b532b9fce8dd8c1f412d297c4b/core/v1/types.go#L1663

File: pkg/apis/app/v1alpha1/application_types.go (hunk 8 above)
Comment (huyhg) [reply pkg/apis/app/v1alpha1/application_types.go]: Similar to above. I'm drawing from the following example in core:
https://github.com/kubernetes/api/blob/184e700b32b7f1b532b9fce8dd8c1f412d297c4b/core/v1/types.go#L1681

File: pkg/apis/app/v1alpha1/application_types.go (hunk 7 above)
Comment (huyhg) [reply pkg/apis/app/v1alpha1/application_types.go]: Done.

File: pkg/apis/app/v1alpha1/application_types.go (hunk 6 above)
Comment (huyhg) [reply pkg/apis/app/v1alpha1/application_types.go]: Changed to `core.ObjectReference`. Discovered that `core.ConfigMapKeySelector` and such also use `LocalObjectReference`, so I created new `ConfigMapKeySelector` and `SecretKeySelector`.

**Comment by huyhg** [comment]
PTAL.

File: pkg/apis/app/v1alpha1/application_types.go [hunk 9]
```diff
@@ -100,6 +101,57 @@ type InfoItem struct {
...
+type ConfigMapKeySelector struct {
+	// The ConfigMap to select from.
+	core.ObjectReference
+	// The key to select.
+	Key string
+}
+
+// SecretKeySelector selects a key from a Secret.
+type SecretKeySelector struct {
+	// The Secrete to select from.
```
Comment (prydonius) [inline pkg/apis/app/v1alpha1/application_types.go]: Secret

**Comment by foxish** [comment]
@kow3ns, can you retry the `/approve`?

**Comment by huyhg** [comment]
Can we merge this change now please?

**Comment by foxish** [comment]
@kow3ns needs /lgtm as well.

**Comment by huyhg** [comment]
@konryd FYI, this has been merged.

**Pull request merged**